core/src/main/java/org/infinispan/conflict/resolution/DefaultConflictResolutionManager.java

```diff
@@ -0,0 +1,159 @@
+import org.infinispan.remoting.rpc.RpcOptions;
+import org.infinispan.remoting.transport.Address;
+
+/**
+ * @author Ryan Emerson
+ */
+public class DefaultConflictResolutionManager<K, V> implements ConflictResolutionManager<K, V> {
+
+   private Cache cache;
+   private CommandsFactory commandsFactory;
+   private DataContainer<K, V> dataContainer;
+   private DistributionManager distributionManager;
+   private StateReceiver<K, V> stateReceiver;
+   private RpcManager rpcManager;
+   private final AtomicBoolean streamInProgress = new AtomicBoolean();
+
+   @Inject
+   public void inject(Cache cache,
+                      CommandsFactory commandsFactory,
+                      DataContainer<K, V> dataContainer,
+                      DistributionManager distributionManager,
+                      StateReceiver<K, V> stateReceiver,
+                      RpcManager rpcManager) {
+      this.cache = cache;
+      this.commandsFactory = commandsFactory;
+      this.dataContainer = dataContainer;
+      this.distributionManager = distributionManager;
+      this.stateReceiver = stateReceiver;
+      this.rpcManager = rpcManager;
+   }
+
+   @Override
+   public Map<Address, InternalCacheValue<V>> getAllVersions(K key) {
+      List<Address> keyReplicaNodes = distributionManager.getConsistentHash().locateOwners(key);
+      Address localAddress = rpcManager.getAddress();
+      Map<Address, InternalCacheValue<V>> versionsMap = new HashMap<>();
+      if (keyReplicaNodes.contains(localAddress)) {
+         InternalCacheValue<V> icv = dataContainer.containsKey(key) ? dataContainer.get(key).toInternalCacheValue() : null;
+         versionsMap.put(localAddress, icv);
+      }
+
+      ClusteredGetCommand cmd = commandsFactory.buildClusteredGetCommand(key, EnumUtil.EMPTY_BIT_SET);
+      long timeout = cache.getCacheConfiguration().clustering().remoteTimeout();
+      RpcOptions rpcOptions = new RpcOptions(timeout, TimeUnit.SECONDS, null, ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS, DeliverOrder.NONE);
+      CompletableFuture<Map<Address, Response>> future = rpcManager.invokeRemotelyAsync(keyReplicaNodes, cmd, rpcOptions);
+      try {
+         Map<Address, Response> responseMap = future.get();
+         for (Map.Entry<Address, Response> entry : responseMap.entrySet()) {
+            if (entry.getValue() instanceof SuccessfulResponse) {
+               SuccessfulResponse response = (SuccessfulResponse) entry.getValue();
+               versionsMap.put(entry.getKey(), (InternalCacheValue<V>) response.getResponseValue());
+            } else {
+               throw new CacheException(String.format("Unable to get key %s from %s: %s", key, entry.getKey(), entry.getValue()));
+            }
+         }
+      } catch (InterruptedException e) {
+         Thread.currentThread().interrupt();
+         throw new CacheException(e);
+      } catch (ExecutionException e) {
+         throw new CacheException(e.getCause());
```

Comment:
Actually if you use `SKIP_OWNERSHIP_CHECK`, it won't throw OTE (if I am not mistaken).